internal/telemetry/telemetry.go

```diff
@@ -69,219 +40,28 @@ func TrackCommand(cmd *cobra.Command) {
-			return
-		}
-
-		event.Properties["auth_method"] = "oauth"
-	}
-}
-
-func withService() eventOpt {
-	return func(event Event) {
-		event.Properties["service"] = config.Service()
-		if config.OpsManagerURL() != "" {
-			event.Properties["ops_manager_url"] = config.OpsManagerURL()
-		}
-	}
-}
-
-func withProjectID(cmd *cobra.Command) eventOpt {
-	return func(event Event) {
-		fromFlag, _ := cmd.Flags().GetString(flag.ProjectID)
-
-		if fromFlag != "" {
-			event.Properties["project_id"] = fromFlag
-			return
-		}
-
-		if config.ProjectID() != "" {
-			event.Properties["project_id"] = config.ProjectID()
-		}
-	}
-}
-
-func withOrgID(cmd *cobra.Command) eventOpt {
-	return func(event Event) {
-		fromFlag, _ := cmd.Flags().GetString(flag.OrgID)
-
-		if fromFlag != "" {
-			event.Properties["org_id"] = fromFlag
-			return
-		}
-
-		if config.ProjectID() != "" {
-			event.Properties["org_id"] = config.OrgID()
-		}
-	}
-}
-
-func withTerminal() eventOpt {
-	return func(event Event) {
-		if cli.IsTerminal(os.Stdout) {
-			event.Properties["terminal"] = "tty"
-			return
-		}
-
-		if cli.IsCygwinTerminal(os.Stdout) {
-			event.Properties["terminal"] = "cygwin"
-		}
-	}
-}
-
-func withInstaller() eventOpt {
-	return func(event Event) {
-		c, err := homebrew.NewChecker(afero.NewOsFs())
-		if err != nil {
-			logError(err)
-			return
-		}
-		if c.IsHomebrew() {
-			event.Properties["installer"] = "brew"
-		}
-	}
-}
-
-func newEvent(opts ...eventOpt) Event {
-	var event = Event{
-		Timestamp: time.Now(),
-		Source:    config.ToolName,
-		Name:      config.ToolName + "-event",
-		Properties: map[string]interface{}{
-			"result": "SUCCESS",
-		},
-	}
-
-	for _, fn := range opts {
-		fn(event)
-	}
-
-	return event
-}
-
-func track(cmd *cobra.Command) {
-	event := newEvent(withCommandPath(cmd), withDuration(cmd), withFlags(cmd), withProfile(), withVersion(), withOS(), withAuthMethod(), withService(), withProjectID(cmd), withOrgID(cmd), withTerminal(), withInstaller())
-
-	cacheDir, err := os.UserCacheDir()
+	t, err := newTracker()
 	if err != nil {
 		logError(err)
 		return
 	}
-	cacheDir = filepath.Join(cacheDir, config.ToolName)
-	err = save(event, cacheDir)
-	if err != nil {
+	if err = t.track(cmd, nil); err != nil {
 		logError(err)
-		return
 	}
 }
 
-func openCacheFile(cacheDir string) (afero.File, error) {
-	exists, err := afero.DirExists(fs, cacheDir)
-	if err != nil {
-		return nil, err
-	}
-	if !exists {
-		if mkdirError := fs.MkdirAll(cacheDir, dirPermissions); mkdirError != nil {
-			return nil, mkdirError
-		}
-	}
-	filename := filepath.Join(cacheDir, cacheFilename)
-	exists, err = afero.Exists(fs, filename)
-	if err != nil {
-		return nil, err
-	}
-	if exists {
-		info, statError := fs.Stat(filename)
-		if statError != nil {
-			return nil, statError
-		}
-		size := info.Size()
-		if size > maxCacheFileSize {
-			return nil, errors.New("telemetry cache file too large")
-		}
+func TrackCommandError(cmd *cobra.Command, e error) {
```

Comment:
The TrackCommand and TrackCommandError functions are almost identical. We can combine them into a single TrackCommand function that takes both parameters, and just call it as appropriate.